### DISCUSSION TITLE
Installing maturing on linux/arm/v7 sometimes uses binary wheel, but other times uses source package

### DISCUSSION THREAD

**Discussion by cuu508** [PyO3/maturin#2384] [Q&A]
I am building a wheel for [the cryptography package](https://github.com/pyca/cryptography) for linux/arm/v7. cryptography is partly written in rust and uses maturin to compile the rust code. When installing the maturin dependency, for some reason, it sometimes finds and uses a wheel, but other times compiles it from source.

Here's a Dockerfile I'm experimenting with:

```
FROM python:3.13.1-slim-bookworm

RUN \
    apt-get update && \
    apt-get install -y build-essential curl libffi-dev libssl-dev libpython3-dev pkg-config
RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | bash -s -- --profile minimal -y
ENV PATH=$PATH:/root/.cargo/bin
RUN \
    pip install --upgrade pip && \
    pip wheel -v --wheel-dir /wheels cryptography
```

And I build it like so:

```
docker buildx build --platform linux/arm/v7 -t sometag --progress plain .
```

In the output I sometimes see:

```
[...]
#7 21.06 Collecting cryptography
#7 21.46   Downloading cryptography-44.0.0.tar.gz (710 kB)
#7 21.67      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 710.7/710.7 kB 6.9 MB/s eta 0:00:00
#7 22.29   Installing build dependencies: started
#7 22.30   Running command pip subprocess to install build dependencies
#7 24.55   Using pip 24.3.1 from /usr/local/lib/python3.13/site-packages/pip (python 3.13)
#7 28.02   Collecting maturin<2,>=1
#7 28.03     Obtaining dependency information for maturin<2,>=1 from https://files.pythonhosted.org/packages/b4/26/f3272ee985ebf9b3e8c4cd4f4efb022af1e12c9f53aed0dcc9a255399f4e/maturin-1.7.8-py3-none-manylinux_2_17_armv7l.manylinux2014_armv7l.musllinux_1_1_armv7l.whl.metadata
#7 28.52     Downloading maturin-1.7.8-py3-none-manylinux_2_17_armv7l.manylinux2014_armv7l.musllinux_1_1_armv7l.whl.metadata (18 kB)
[...]
#7 55.15   Downloading maturin-1.7.8-py3-none-manylinux_2_17_armv7l.manylinux2014_armv7l.musllinux_1_1_armv7l.whl (7.4 MB)
[...]
```

But other times:

```
[...]
#7 19.35 Collecting cryptography
#7 19.80   Downloading cryptography-44.0.0.tar.gz (710 kB)
#7 20.03      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 710.7/710.7 kB 4.9 MB/s eta 0:00:00
#7 20.62   Installing build dependencies: started
#7 20.62   Running command pip subprocess to install build dependencies
#7 22.95   Using pip 24.3.1 from /usr/local/lib/python3.13/site-packages/pip (python 3.13)
#7 25.77   Collecting maturin<2,>=1
#7 26.26     Downloading maturin-1.7.8.tar.gz (195 kB)
[...]
```

This is a problem for me because compiling maturin from source for linux/arm/v7, using QEMU emulation is slow. On my local system (Intel 9700K, 8 cores):

* when maturin is installed from wheel, cryptography builds in 6.5 minutes
* when maturin is compiled from source, cryptography builds in 26 minutes

But it's worse when running in GitHub Actions (4 cores, not sure what CPU). The project I'm working on sometimes builds in ~40 minutes, but other times times out after 6 hours. I'm not 100% sure the 6 hours are spent building maturin, but I suspect it's at least a contributing factor.

Any tips what is going here, why does pip find and use a wheel only sometimes?
Thanks!

**Comment by messense** [comment]
Hard to say but you can debug it futher by using `pip install -vv` to get a verboser output.
